### PR TITLE
GCI-427: Check payment with payments.api when patching order

### PR DESCRIPTION
**Pull request by jphillips2121**
Implements callback to the payments api when the patch endpoint on the payment-details URL has been hit. 

This verifies the payment exists, is paid, is the same amount as the checkout, and has the correct url as a resource before marking the checkout as paid. 

If the incoming patch request does not have a "paid" status the checkout is instead changed the incoming status, and the basket is not cleared.

Resolves: GCI-427 

Related PR's for story: 
https://github.com/companieshouse/api-sdk-manager-java-library/pull/7
https://github.com/companieshouse/payments.api.ch.gov.uk/pull/124

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java [hunk 1]
```diff
@@ -229,6 +288,17 @@ private void processSuccessfulPayment(final String requestId,
         trace("Cleared basket: " + basket, requestId);
     }
 
+    private Double calculateTotalAmountToBePaid(Checkout checkout) {
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:299]: I note other calculations of amounts in the Orders API have used integer values.

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java [hunk 2]
```diff
@@ -195,25 +197,82 @@ public BasketController(final BasketMapper mapper,
...
+        } else {
+
+            // Update the checkout of non-paid order
+            updateCheckout(checkout, basketPaymentRequestDTO);
         }
-        return ResponseEntity.ok("");
+
+        trace("Order " + id + " has been marked as " + basketPaymentRequestDTO.getStatus() + " with payment id " + basketPaymentRequestDTO.getPaymentReference(), requestId);
+
+        return ResponseEntity.status(NO_CONTENT).body(null);
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:266]: What is the significance of this change to the return value?

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java [hunk 3]
```diff
@@ -195,25 +197,82 @@ public BasketController(final BasketMapper mapper,
...
+                return ResponseEntity.status(BAD_REQUEST).body("Payment is not set to paid in payment api for payment " + basketPaymentRequestDTO.getPaymentReference());
+            }
+
+            // Check the amount paid in the payment session and the amount expected in the order are the same
+            if (Double.parseDouble(paymentSession.getAmount()) != calculateTotalAmountToBePaid(checkout)) {
+                LOGGER.error("Total amount paid for with payment session " + basketPaymentRequestDTO.getPaymentReference() + ": " + paymentSession.getAmount()
+                        + " does not match amount expected for order " + id + ": " + checkoutData.getTotalOrderCost());
+                return ResponseEntity.status(BAD_REQUEST).body("Total amount paid for with payment session " + basketPaymentRequestDTO.getPaymentReference() + ": "
+                        + paymentSession.getAmount() + " does not match amount expected for order " + id + ": " + checkoutData.getTotalOrderCost());
+            }
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:237]: Message logged and response body are the same. Easier to maintain and read if the message is in a variable?

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java [hunk 4]
```diff
@@ -195,25 +197,82 @@ public BasketController(final BasketMapper mapper,
...
+            // Get the URI for the resource in the payments session
+            String paymentsResourceUri = paymentSession.getLinks().get("resource")
+                    .substring(paymentSession.getLinks().get("resource").lastIndexOf("/basket/checkouts/"));
+            // Check that the URI that has been requested to mark as paid, matches URI from the payments session
+            if (!paymentsResourceUri.equals(request.getRequestURI())) {
+                LOGGER.error("The URI that is attempted to be closed " + request.getRequestURI()
+                        + " does not match the URI that the payment session is created for " + paymentsResourceUri);
+                return ResponseEntity.status(BAD_REQUEST).body("The URI that is attempted to be closed " + request.getRequestURI()
+                        + " does not match the URI that the payment session is created for " + paymentsResourceUri);
+            }
```
Comment (VaughanJackson123) [inline src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:248]: Message logged and response body are the same. Easier to maintain and read if the message is in a variable?

File: src/test/java/uk/gov/companieshouse/orders/api/controller/BasketControllerIntegrationTest.java [hunk 5]
```diff
@@ -574,344 +578,217 @@ public void patchBasketReturnsBadRequestItemUriInvalid() throws Exception {
...
-        checkout.setId(CHECKOUT_ID);
-        checkoutRepository.save(checkout);
-
-        BasketPaymentRequestDTO basketPaymentRequestDTO = new BasketPaymentRequestDTO();
-        basketPaymentRequestDTO.setPaidAt(LocalDateTime.now());
-        basketPaymentRequestDTO.setPaymentReference("reference");
-        basketPaymentRequestDTO.setStatus(PaymentStatus.PAID);
-
-        timestamps.start();
+    @DisplayName("Patch basket payment details failure checkout id doesn't not exist")
```
Comment (VaughanJackson123) [inline src/test/java/uk/gov/companieshouse/orders/api/controller/BasketControllerIntegrationTest.java:675]: Not sure I understand the content of the display name.

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java (hunk 1 above)
Comment (jphillips2121) [reply src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:299]: The payments.api returns the total amount for the payment in decimal format `70.00` so the amount we are calculating here will also need to be in decimal format. We could calculate it as an integer and then convert it to decimal format but this seemed more succinct. 

Is there no possibility of an order having a decimal payment? Like `£31.50` ?

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java (hunk 2 above)
Comment (jphillips2121) [reply src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:266]: The payment processed consumer [expects a NO_CONTENT status](https://github.com/companieshouse/payment-processed-consumer/blob/develop/payment/payment.go#L84) to be returned. If anything else is returned it retries the call over and over

File: src/test/java/uk/gov/companieshouse/orders/api/controller/BasketControllerIntegrationTest.java (hunk 5 above)
Comment (jphillips2121) [reply src/test/java/uk/gov/companieshouse/orders/api/controller/BasketControllerIntegrationTest.java]: I've updated the display names to hopefully make a bit more sense

File: src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java (hunk 1 above)
Comment (VaughanJackson123) [reply src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:299]: Apparently to date that possibility has not arisen. It was discussed a long time ago and ruled  out at that stage. From what you're saying, this is effectively a contractual requirement that we cannot avoid. If that's what works then so be it. Perhaps a comment noting that in the code would be good.

Comment (jphillips2121) [reply src/main/java/uk/gov/companieshouse/orders/api/controller/BasketController.java:304]: Done 👍

**Pull request merged**